docs/design/matrix.rst

```diff
@@ -0,0 +1,167 @@
+    uniform matrix<uniform unsigned int8, M, N> A; // error, uniform/varying qualifiers are not allowed with matrix types
+    varying matrix<uniform unsigned int8, M, N> A; // error, uniform/varying qualifiers are not allowed with matrix types
+
+
+Basic operators
+---------------
+To access the particular matrix element[s], the standard matrix access operator can be used: `A[X, Y]`, where X, Y are of integer type.
+
+If `X` and `Y` are uniform, the result is a single matrix element.
+
+If `X` or `Y` are varying, the result is a `varying`.
+
+If `X` or `Y` are both varying, the result is a new matrix of size `programCount` x `programCount`.
+
+If the index is varying and `programCount` is bigger than matrix dimension, ISPC will produce an error.
+
+Consider an example:
+
+.. code-block:: cpp
+
+    /*
+    Matrix A:
+    a11 | a12 | a13 | a14 | a15 | a16 | a17 | a18
+    a21 | a22 | a23 | a24 | a25 | a26 | a27 | a28
+    a31 | a32 | a33 | a34 | a35 | a36 | a37 | a38
+    a41 | a42 | a43 | a44 | a45 | a46 | a47 | a48
+    */
+    uniform int X = 1;
+    uniform int Y = 2;
+    uniform int xy = A[X, Y]; // returns a23
+
+    varying int X = 1;
+    uniform int Y = 2;
+    varying int xy = A[X, Y]; // returns {a23, a23, a23, a23}
+
+    uniform int X = 2;
+    varying int Y = {2, 1, 3, 5};
+    varying int xy = A[X, Y]; // returns {a33, a32, a34, a36}
+
+    varying int X = {0, 1, 2, 3};
+    varying int Y = {2, 1, 3, 5};
+    // Returns
+    // a13 | a11 | a14 | a16
+    // a23 | a21 | a24 | a26
+    // a33 | a31 | a34 | a36
+    // a43 | a41 | a44 | a46
+    matrix<int, programCount, programCount> XY = A[X, Y];
+
+
+The following ISPC operators are supported for matrices of the same types or matrix and a scalar:
+
+.. list-table:: Operators
+
+  * - Symbols
+    - Use
+  * - ``=``
+    - Assignment
+  * - ``+``, ``-``
+    - Element-wise addition and subtraction
+  * - ``++``, ``--``
+    - Element-wise increment/decrement
+  * - ``+=``, ``-=``
+    - Compound assignment operators
+
+Operators ``+``, ``-``, ``*``, ``/`` can be used with matrix and uniform value, applying the operation to each element of the matrix.
+
+Examples:
+
+.. code-block:: cpp
+
+    matrix<uniform int, M, N> A = 1; // init all matrix elements with 1
+    matrix<uniform int, M, N> B = A + 2; // add 2 to each element of matrix A and init with result matrix B
+
+Conversions
+-----------
+Matrix type can be converted to another matrix type if the number of rows and columns are the same and the value's elements can be converted to the element type of the result type.
+
+.. code-block:: cpp
+
+    matrix<uniform int16, M, N> A16;
+    matrix<uniform float, M, N> AF = (matrix<uniform float, M, N>)A16;
+
+Interoperability
+----------------
+Matrix is internal ISPC type. It can't be used as an argument to `export` or `extern "C"` functions. It can be used as an argument for internal ISPC functions.
```

Comment:
Does that mean that C/C++ code will not know anything about layout/structure/restrictions of ISPC matrix types? I wonder how matrix initialization C/C++ code could look like.

Reply:
Right. On C/C++side it may look like this:
```
    template <class T> void init_matrix(std::vector<T> &M, unsigned int rows, unsigned int cols, T value) {
    for (unsigned int r = 0; r < rows; r++)
        for (unsigned int c = 0; c < cols; c++) {
            M[r * cols + c] = value;
        }
    }
    int main() {
    ...
    std::vector<SRC_T> matrixA(M * N);
    std::vector<SRC_T> matrixB(N * K);
    std::vector<DST_T> matrixC(M * K);
    init_matrix<SRC_T>(matrixA, M, N, 1);
    init_matrix<SRC_T>(matrixB, N, K, 0.5);
    init_matrix<DST_T>(matrixC, M, K, 0);
    
    // If ISPCRT is used:
    ispcrt::Device device(ISPCRT_DEVICE_TYPE_GPU);
    ispcrt::Array<unsigned> matrixA_dev(device, matrixA);
    ispcrt::Array<unsigned> matrixB_dev(device, matrixB);
    ispcrt::Array<DST_T> matrixC_dev(device, matrixC);

    // Setup parameters structure
    Parameters<DST_T> p;

    p.mA = matrixA_dev.devicePtr();
    p.mB = matrixB_dev.devicePtr();
    p.mC = matrixC_dev.devicePtr();
    p.M = M;
    p.N = N;
    p.K = K;
    auto p_dev = ispcrt::Array<Parameters<DST_T>>(device, p);
    // Create module and kernel
   ...
    // Create task queue and execute kernel
    ispcrt::TaskQueue queue(device);

    queue.copyToDevice(p_dev);
    queue.copyToDevice(matrixA_dev);
    queue.copyToDevice(matrixB_dev);
    queue.copyToDevice(matrixC_dev);
   ...
}
```